### DISCUSSION TITLE
typed and optimization

### DISCUSSION THREAD

**Discussion by kusumotonorio** [factor/factor#2428] [General]
Do words defined with `TYPED:` tell the compiler at compile time the class information of the value it will put on the stack to optimize the words it concatenates?
It appears to be the case. How exactly does it do that?

I'm looking for a way to eliminate dispatching of the later word as much as possible when concatenating multi-generic words. I could do it like the word defined in `TYPED:`, but I suspect that inspecting the output would slow it down. I only want to make the multi-generic word tell the class information of the output.

**Comment by mrjbq7** [comment]
Typed words have two layers. 

An outer word that checks the stack input type and throws an exception if the types are not correct at run-time.  The compiler removes or reduces these checks if type information is known. 

An inner word which declares the types on the stack (for purposes of more efficient code generation in the inner word). This uses the kernel.private:declare word to do so. 

If all types are known at compile time then only the inner word is being called, which is efficient.

**Reply by kusumotonorio** [reply]
I added `declare` to the end of the definition for multi-generic words where all methods have the same class group output.
(Multi-methods with a class specification of output will have their output checked by the feature of `typed` .)

```factor
MGENERIC: float+float ( a b -- c )
MM: float+float ( a: float b: float -- c: float ) float+ ;
MM: float+float ( a: fixnum b: fixnum -- c: float ) fixnum+ >float ;
```
```factor
IN: scratchpad \ float+float def>> .
[
    dup float? [ over float? ] [ f ] if
    [ MM\ float+float ( a: float b: float -- c: float ) ] [
        dup fixnum? [ over fixnum? ] [ f ] if
        [ MM\ float+float ( a: fixnum b: fixnum -- c: float ) ]
        [ 2 \ float+float [ narray ] dip no-method ] if
    ] if { float } declare
]
IN: scratchpad 
```

Apparently it works as optimization information for the next word. Is such a usage example correct? It's surprising because I thought `declare` needed to be interrupted  in before the next word.

**Comment by kusumotonorio** [comment]
As for how much of the type information on the stack is used by factor's optimizing compiler, apparently not as much as I had imagined.
The information given in `declare` appears to be used only for the word immediately following it.

It would be nice to receive more information if `{ float } declare` was followed by a literal of `fixnum` on the stack, so that the information would be equivalent to `{ float fixnum } declare`, or even `swap dup` afterwards, so that the information would be equivalent to `{ fixnum float float } declare`.

If so, multi-generic words would have a better chance of removing dispatch code, and would be faster. It is faster to be able to remove the dispatch than any fast dispatch algorithm.
